### PR TITLE
Improve 2D simulation example

### PR DESCRIPTION
**Pull request by ja-che**

### CODE REVIEW

File: examples/plot_2D_simulation_example.py [hunk 1]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+correction_no_cluster = 1. / n_features
+correction_cluster = 1. / n_clusters
+thr_c = zscore_from_pval((fwer_target / 2) * correction_cluster)
+thr_nc = zscore_from_pval((fwer_target / 2) * correction_no_cluster)
+
+#############################################################################
+# Inference with several algorithms
+# ---------------------------------
+
+#############################################################################
```
Comment (agramfort) [inline examples/plot_2D_simulation_example.py:226]:
```suggestion
#############################################################################
# Inference with several algorithms
# ---------------------------------
#
```

File: examples/plot_2D_simulation_example.py [hunk 2]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+zscore = zscore_from_pval(pval, one_minus_pval)
+selected_dl = zscore > thr_nc  # use the "no clustering threshold"
+
+# compute estimated support (second method)
+selected_dl = np.logical_or(pval_corr < fwer_target / 2,
+                            one_minus_pval_corr < fwer_target / 2)
+
+#############################################################################
+# Now, we compute the support estimated using a clustered inference algorithm
+# (c.f. References) called Clustered Desparsified Lasso (CluDL) since it
```
Comment (agramfort) [inline examples/plot_2D_simulation_example.py:256]: can References be link to it?

File: examples/plot_2D_simulation_example.py [hunk 3]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+zscore = zscore_from_pval(pval, one_minus_pval)
+selected_cdl = zscore > thr_c  # use the "clustering threshold"
+
+# compute estimated support (second method)
+selected_cdl = np.logical_or(pval_corr < fwer_target / 2,
+                             one_minus_pval_corr < fwer_target / 2)
+
+#############################################################################
+# Finally, we compute the support estimated by an ensembled clustered
+# inference algorithm (c.f. References). This algorithm is called EnCluDL
```
Comment (agramfort) [inline examples/plot_2D_simulation_example.py:280]: idem

File: examples/plot_2D_simulation_example.py [hunk 4]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+
+maps.append(np.reshape(selected_ecdl, shape))
+titles.append('EnCluDL')
+
+plot(maps, titles)
+
+#############################################################################
+# Analysis of the results
+# -----------------------
+# As advocated in introduction, the standard method that do not compress the
```
Comment (agramfort) [inline examples/plot_2D_simulation_example.py:334]: introduction?

**Comment by agramfort** [comment]
rendered doc https://99-259908992-gh.circle-artifacts.com/0/dev/auto_examples/plot_2D_simulation_example.html#sphx-glr-auto-examples-plot-2d-simulation-example-py

File: examples/plot_2D_simulation_example.py [hunk 5]
```diff
@@ -3,8 +3,56 @@
...
 
+This example shows the advantages of spatially relaxed inference when
+dealing with high-dimensional spatial data. To do so, we compare several
+statistical methods that aim at recovering the support, i.e., predictive
+features. Among those methods some leverage the spatial structure of the
+data. For more details about the inference algorithms presented in this
+example or about the generative process used to simulate the data,
+please refer to Chevalier et al. (2021) [1]_.
+
+This example corresponds exactly to the experiment thorougly described in
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:14]:
```suggestion
This example corresponds  to the experiment described in detail in
```

File: examples/plot_2D_simulation_example.py [hunk 6]
```diff
@@ -3,8 +3,56 @@
...
+Chevalier et al. (2021) [1]_. Shortly, to simulate the data, we draw
+``n_samples`` i.i.d Gaussian vectors of size ``n_features`` and reshape them
+into squares (edges are equal to ``n_features ** (1/2)``). Then, to introduce
+some spatial structure, we apply a Gaussian filter that correlates features
+that are nearby. The 2D data are then flattened into a design matrix ``X`` to
+recover the standard machine learning quantity and to ease the computation of
+the simulated target ``y`` (see below). Then, we construct the weight map ``w``
+which displays the same shape as the 2D data, it contains four predictive
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:22]:
```suggestion
which has the same shape as the 2D data, as it contains four predictive
```

File: examples/plot_2D_simulation_example.py [hunk 7]
```diff
@@ -3,8 +3,56 @@
...
+example or about the generative process used to simulate the data,
+please refer to Chevalier et al. (2021) [1]_.
+
+This example corresponds exactly to the experiment thorougly described in
+Chevalier et al. (2021) [1]_. Shortly, to simulate the data, we draw
+``n_samples`` i.i.d Gaussian vectors of size ``n_features`` and reshape them
+into squares (edges are equal to ``n_features ** (1/2)``). Then, to introduce
+some spatial structure, we apply a Gaussian filter that correlates features
+that are nearby. The 2D data are then flattened into a design matrix ``X`` to
+recover the standard machine learning quantity and to ease the computation of
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:20]:
```suggestion
represent it as a regression setting and to ease the computation of
```

File: examples/plot_2D_simulation_example.py [hunk 8]
```diff
@@ -3,8 +3,56 @@
...
+of ``X``, the map ``w`` is finally flattened into a vector ``beta``. Lastly,
+to derive the target ``y``, we draw a white Gaussian noise ``epsilon`` and
+use a linear generative model: ``y = X beta + epsilon``.
+
+The results of this experiment show that the methods that leverage the spatial
+structure of the data are relevant. More precisely, we show that clustered
+inference algorithms (e.g., CluDL) and ensembled clustered inference algorithms
+(e.g., EnCluDL) are more powerful than the standard inference methods (see also
+Chevalier et al. (2021) [1]_). Indeed, when the number of features is much
+greater than the number of samples, the standard statistical methods are
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:33]:
```suggestion
greater than the number of samples, standard statistical methods are
```

File: examples/plot_2D_simulation_example.py [hunk 9]
```diff
@@ -3,8 +3,56 @@
...
+unlikely to recover the support. Then, the idea of clustered inference is to
+compress the data without breaking the spatial structure, leading to a
+compressed problem (very) close to the original problem. This enables a
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:36]:
```suggestion
compressed problem close to the original problem. This leads to a
```

File: examples/plot_2D_simulation_example.py [hunk 10]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+# computation parameter
+n_jobs = 1
+
+##############################################################################
+# Computing z-score thresholds for support estimation
+# ---------------------------------------------------
+#
+# Below, we translate the FWER target into z-score targets.
+# To compute the z-score targets we also take into account for the multiple
+# testing correction, to do so we consider the Bonferroni correction.
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:209]:
```suggestion
# testing correction. To do so we consider the Bonferroni correction.
```

File: examples/plot_2D_simulation_example.py [hunk 11]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+# uncertainty). Formally, "spatial tolerance" is defined by the largest
+# distance from the true support for which the occurence of a false discovery
+# is not statistically controlled (c.f. References).
+# Theoretically, the spatial tolerance ``delta`` is equal to the largest
+# cluster diameter. However this choice is conservative, notably in the case
+# of ensembled clustered inference. For these algorithms, we recommend to take
+# the average cluster radius. In this example, we choose ``n_clusters = 200``,
+# leading to a theoretical spatial tolerance ``delta = 6``. However
+# following our pratical guidance ``delta = 2`` would have sufficient
+# for ensembled clustered inference algorithms (see Results).
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:191]: This last sentence is unclear. WDYM with "our practical guidance" ? Simply, "it turns out that..."

File: examples/plot_2D_simulation_example.py [hunk 12]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+
+#############################################################################
+# Finally, we compute the support estimated by an ensembled clustered
+# inference algorithm (c.f. References). This algorithm is called EnCluDL
+# Ensemble of Clustered Desparsified Lasso (EnCluDL) since it runs several
+# CluDL algorithms with different clustering choices. The different CluDL
+# solutions are then aggregated into one.
+
+# define the FeatureAgglomeration object that performs the clustering
+connectivity = image.grid_to_graph(n_x=shape[0],
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:286]: Why do you recompute connectivity ?

File: examples/plot_2D_simulation_example.py [hunk 13]
```diff
@@ -84,97 +150,194 @@ def plot(maps, titles, save_fig=False):
...
+                                   n_y=shape[1])
+ward = FeatureAgglomeration(n_clusters=n_clusters,
```
Comment (bthirion) [inline examples/plot_2D_simulation_example.py:288]: you don't need to redefine ward.

**Review by bthirion — COMMENTED** [review]
Thx for te pedagogical efforts !

**Comment by ja-che** [comment]
I have made the requested modifications. To me, it is ready to merge !

rendered doc: https://100-259908992-gh.circle-artifacts.com/0/dev/auto_examples/plot_2D_simulation_example.html#sphx-glr-auto-examples-plot-2d-simulation-example-py

**Comment by bthirion** [comment]
Thx !

**Pull request merged**